Payment Requests

Kukks opened the pull request:
With this PR, I am introducing a new entity to BtcPay, Payment Requests as discussed in #314 

Closes #314 
Also Closes #571

A payment request is exactly what it sounds, a request to be paid. It consists of a title, description, amount and an optional expiry date. 

Until it expires, it can be paid at any time by creating a fresh invoice of the remaining balance on demand. Only one invoice can be active at a time per payment request.

It comes with the same real-time updates as the crowdfunding but with less colorful aspects to it. :) 

The UI will also display all invoices created through it along with all payment, tx ids and block explorer links. 

I will also add a print-friendly version that can be used to print for accounting purposes. 

Eventually, I will build #461  on top of Payment Requests since a payment request is perfect to send to an email on a schedule.


![image](https://user-images.githubusercontent.com/1818366/51250223-03bcbb80-1996-11e9-9bf4-484ff6e5bed8.png)

Kukks commented:
Added UTs, ready for you @NicolasDorier  💃

NicolasDorier commented on BTCPayServer/Controllers/AppsPublicController.cs:
```diff
@@ -428,6 +431,62 @@ public ViewPointOfSaleViewModel.Item[] Parse(string template, string currency)
+
+         public Dictionary<string, decimal> GetCurrentContributionAmountStats(InvoiceEntity[] invoices, bool usePaymentData = true)
+        {
+            if(usePaymentData){
```
format

NicolasDorier commented on BTCPayServer/Migrations/20190116102421_AddPaymentRequests.Designer.cs:
```diff
@@ -0,0 +1,602 @@
+using Microsoft.EntityFrameworkCore.Migrations;
+using Microsoft.EntityFrameworkCore.Storage.ValueConversion;
+
+namespace BTCPayServer.Migrations
```
What is this file? I don't remember  any file like this needed to be generated by entity when changing the DB.

Kukks replied:
It was generated by the .net cli ef core tool. I dare not touch these :)

NicolasDorier commented on BTCPayServer/PaymentRequest/PaymentRequestHub.cs:
```diff
@@ -0,0 +1,211 @@
+                    Amount = entity.ProductInformation.Price,
+                    Currency = entity.ProductInformation.Currency,
+                    ExpiryDate = entity.ExpirationTime.DateTime,
+                    Status = Enum.GetName(typeof(InvoiceStatus), entity.Status),
```
Don't use this, there is a `GetState()` I think that now you can use which indicate both the status and the exceptionStatus properly.

NicolasDorier commented on BTCPayServer/Services/PaymentRequests/PaymentRequestRepository.cs:
```diff
@@ -0,0 +1,196 @@
+    {
+        private readonly ApplicationDbContextFactory _ContextFactory;
+        private readonly InvoiceRepository _InvoiceRepository;
+        private readonly EventAggregator _EventAggregator;
```
No event aggregator in the Repository, this is not the goal. A repository just abstract the storage technology.

Kukks commented:
PR changes done @NicolasDorier deet

Kukks commented:
don't forget to review or I will just keep adding stuff @NicolasDorier 

 I refactored the status to make the hosted service lighter

NicolasDorier commented on BTCPayServer/Migrations/20190121133309_AddPaymentRequests.Designer.cs:
```diff
@@ -0,0 +1,604 @@
+﻿// <auto-generated />
```
What is this file? it just rebuild the whole Models of all table Oo

NicolasDorier replied:
I think you can remove it

NicolasDorier commented on BTCPayServer/Controllers/AppsPublicController.cs:
```diff
@@ -428,6 +428,62 @@ public ViewPointOfSaleViewModel.Item[] Parse(string template, string currency)
+                        if (string.Equals(PaymentMethodId.Parse(stat.Key).CryptoCode, rateTask.Key.Right,
+                            StringComparison.InvariantCultureIgnoreCase))
+                        {
+                            result += (1m / rate.Value) *  stat.Value;
```
bug, additions are not atomic.
You are over complicating, you don't need `finalTasks` at all.

NicolasDorier commented on BTCPayServer/Controllers/PaymentRequestController.cs:
```diff
@@ -0,0 +1,315 @@
+        }
+
+        [HttpPost]
+        [Route("remove/{id}")]
```
can you instead `{id}/remove` ?

NicolasDorier commented on BTCPayServer/Controllers/PaymentRequestController.cs:
```diff
@@ -0,0 +1,315 @@
+        }
+
+        [HttpGet]
+        [Authorize(AuthenticationSchemes = Policies.CookieAuthentication)]
```
Put the authorize at the `Controller` level.

NicolasDorier commented on BTCPayServer/Services/PaymentRequests/PaymentRequestRepository.cs:
```diff
@@ -0,0 +1,249 @@
+        {
+            var result = Blob == null
+                ? new PaymentRequestBlob()
+                : new Serializer(Network.Main).ToObject<PaymentRequestBlob>(Encoding.UTF8.GetString(Blob));
```
Don't hard code `Network.Main`. Why do you need the serializer at all? It seems to me you can use JsonConvert directly.

Kukks commented:
ready for another round @NicolasDorier  :P

NicolasDorier commented:
btw, let's keep it like that, but in the future I would prefer if JS was not mandatory, and instead just used to improve the user experience if available.
The apps are different.

Kukks commented:
I can do it, it's not a big thing. Give me an hour

Kukks commented:
Added No js version @NicolasDorier

britttttk commented:
This is an awesome feature, I'm super exited for this one.
My comments:

1. **expiry** - I've never heard this word used before in the US :P usually expiration 
2. **Calendar**:
    - What is the purpose of allowing users to choose past dates, if there is an option for no expiration? I don't think it can be paid if it's created as expired?
    - Is it possible, when the user re-opens the calendar for it to open to the date they have chosen?
    - Is it possible for calendar info to be kept through a PayReq error? I mean if they choose Bitcoin instead of BTC, the error deletes their calendar info and they have to do it again.
3. **No Deriv** - Causes the 'Make Payment' button on the PayReq page to hang forever... would maybe be better if it does nothing when pressed (how it is for CF App).
4. **Currency Choice**:
![2currencytypes](https://user-images.githubusercontent.com/39231115/51872669-868b3200-2317-11e9-9a4f-7a241c091a37.PNG)
    - What is the correct option here? One choice? A list?
    - The error message is very strange requesting a string or an array type? BTC, LTC doesn't work.
    - The error message displays incorrectly: "The Field The Currency..."
5. **UI Dashes:** - Normally there are dashes here
![dashes](https://user-images.githubusercontent.com/39231115/51872835-0e713c00-2318-11e9-8f1d-563f2479ca09.PNG)

Kukks commented:
resolved all above ^

britttttk commented:
@Kukks Thanks for fixing everything. 

Should the Currency Field* be starred to show it's required? The light BTC text is great but if the user assumes it's pre-typed for you, the page hangs when they try to save? It's giving me a null exception error in VS. I'm not sure how that appears to the user. Is it ok?

NicolasDorier commented:
there is still some `.designer`

NicolasDorier commented on BTCPayServer/Controllers/AppsPublicController.cs:
```diff
@@ -441,6 +441,62 @@ public ViewPointOfSaleViewModel.Item[] Parse(string template, string currency)
                 .ToArray();
         }
 
+        public async Task<decimal> GetCurrentContributionAmount(Dictionary<string, decimal> stats, string primaryCurrency, RateRules rateRules)
```
Why are you making it so complicated?

```csharp
        private async Task<decimal> GetCurrentContributionAmount(Dictionary<string, decimal> stats, string primaryCurrency, RateRules rateRules)
        {
            decimal result = 0;

            var ratesTask = _RateFetcher .FetchRates(
                stats.Keys
                    .Select((x) => new CurrencyPair( primaryCurrency, PaymentMethodId.Parse(x).CryptoCode))
                    .Distinct()
                    .ToHashSet(), 
                rateRules);
            foreach (var rateTask in ratesTask)
            {
                    var tResult = await rateTask.Value;
                    var rate = tResult.BidAsk?.Bid;
                    if (rate == null) continue;

                    foreach (var stat in stats)
                    {
                        if (string.Equals(PaymentMethodId.Parse(stat.Key).CryptoCode, rateTask.Key.Right,
                            StringComparison.InvariantCultureIgnoreCase))
                        {
                            result += (1m / rate.Value) *  stat.Value;
                        }
                    }
            }
            return result;
        }
```

Kukks replied:
Simplified it. this thing is used in crowdfund app too

NicolasDorier replied:
yes, and it is probably bugged as it update result concurrently.

Kukks commented:
Ok @NicolasDorier  & @britttttk  

I added in server-side formatted amount like you wanted @NicolasDorier  and added * next to the required fields and fixed null error on invalid currency input.

Are you all done :P ?

britttttk commented:
@Kukks Thanks for fixing ^

The Amount field does not actually require user input (as you now denote with the asterisk).

As a result, the PayReq defaults to 0 which does not cause any problems in BTCPay. But what is the purpose of allowing PayReqAmt = 0 if it can't be paid? Only viewed, I can't imagine a use case for this. Some accounting reason perhaps, but it might be better to wait for someone to request this, imo. 

![image](https://user-images.githubusercontent.com/39231115/52010804-a5163800-2493-11e9-8926-f1f0b596e3c7.png)

Specifically, what purpose does it serve for a user to create a PayReq which appears like this:

![image](https://user-images.githubusercontent.com/39231115/52010883-d858c700-2493-11e9-804b-34b2bf119b72.png)

Kukks commented:
It's more work to "remove" the feature so let's not. While in creating
state, it is fine to have 0 in it anyway

On Wed, Jan 30, 2019, 21:39 britttttk <notifications@github.com wrote:

> @Kukks <https://github.com/Kukks> Thanks for fixing ^
>
> The Amount field does not actually require user input (as you now denote
> with the asterisk).
>
> As a result, the PayReq defaults to 0 which does not cause any problems in
> BTCPay. But what is the purpose of allowing PayReqAmt = 0 if it can't be
> paid? Only viewed, I can't imagine a use case for this. Some accounting
> reason perhaps, but it might be better to wait for someone to request this,
> imo.
>
> [image: image]
> <https://user-images.githubusercontent.com/39231115/52010804-a5163800-2493-11e9-8926-f1f0b596e3c7.png>
>
> Specifically, what purpose does it serve for a user to create a PayReq
> which appears like this:
>
> [image: image]
> <https://user-images.githubusercontent.com/39231115/52010883-d858c700-2493-11e9-804b-34b2bf119b72.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/btcpayserver/btcpayserver/pull/532#issuecomment-459100187>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABu-_lwjmejRwPZapsl-fMdDbidmXvrWks5vIgL-gaJpZM4aBxXx>
> .
>

britttttk commented:
tACK 

Thanks for your work @Kukks this is a really cool feature :D

NicolasDorier commented on BTCPayServer/Views/PaymentRequest/MinimalPaymentRequest.cshtml:
```diff
@@ -0,0 +1,161 @@
+            @Model.Title
+
+            <span class="text-muted float-right text-center">
+                @switch (status)
```
Can you put this code in the controller instead? Adding logic in the View really suck.

Kukks replied:
(ノಠ ∩ಠ)ノ彡( \o°o)\

NicolasDorier replied:
I will do it ok ok :D

NicolasDorier replied:
But generally, logic in razor is bitch:
* Refactors does like rename does not change impact razor.cshtml
* It crashes at runtime

pavlenex commented:
This is a great PR with plenty of potential use-cases. It proved to be way better than I originally suggested.

My suggestions are more on the front-end side:

1. Rename Payment Request to Bills (BTCPay Bills sounds great), unless there is additional use-case beyond billing that you'd like to implement.

2. Remove text editor and use simple text field. (Text editor proved to be buggy sometimes. In billing, we don't need to add videos images etc.

3. Center the paymenr request back end (I did simple margin: auto;) and it looks much more focused.

![margins](https://user-images.githubusercontent.com/36959754/52163364-30383d80-26e1-11e9-84b6-6d43e115c837.PNG)

4. Move Store ID field bellow currency, it has more context there

5. Save settings and checkbox are not that good UX. Can we have "Save for Later" and ("Create Bill" or "Publish") without the checkbox. It does not make much sense and people will in most cases forget it, just like we do with the Crowfund app, so many people forget to check that box.

6. Add padding on top of the payment request so that it looks centered on the page and more focused.
![capture44](https://user-images.githubusercontent.com/36959754/52163193-27df0300-26df-11e9-82eb-fa15abfb0ea3.PNG)

7. Make a payment button, should be **Pay** or Pay Now.

Kukks commented:
I think the text editor should stay. Too many cool use cases can come out
of it.




On Sat, Feb 2, 2019, 11:55 Pavle <notifications@github.com wrote:

> This is a great PR with plenty of potential use-cases. It proved to be way
> better than I originally suggested.
>
> My suggestions are more on the front-end side:
>
>    1.
>
>    Rename Payment Request to Bills (BTCPay Bills sounds great), unless
>    there is additional use-case beyond billing that you'd like to implement.
>    2.
>
>    Remove text editor and use simple text field. (Text editor proved to
>    be buggy sometimes. In billing, we don't need to add videos images etc.
>    3.
>
>    Center the paymenr request back end (I did simple margin: auto;) and
>    it looks much more focused.
>
> [image: margins]
> <https://user-images.githubusercontent.com/36959754/52163364-30383d80-26e1-11e9-84b6-6d43e115c837.PNG>
>
>    1.
>
>    Move Store ID field bellow currency, it has more context there
>    2.
>
>    Save settings and checkbox are not that good UX. Can we have "Save for
>    Later" and ("Create Bill" or "Publish") without the checkbox. It does not
>    make much sense and people will in most cases forget it, just like we do
>    with the Crowfund app, so many people forget to check that box.
>    3.
>
>    Add padding on top of the payment request so that it looks centered on
>    the page and more focused.
>    [image: capture44]
>    <https://user-images.githubusercontent.com/36959754/52163193-27df0300-26df-11e9-82eb-fa15abfb0ea3.PNG>
>    4.
>
>    Make a payment button, should be *Pay* or Pay Now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/btcpayserver/btcpayserver/pull/532#issuecomment-459955720>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABu-_hCmHC67NnfVpfSTfqT-bzMBmzwSks5vJW67gaJpZM4aBxXx>
> .
>

britttttk commented:
> Remove text editor

I think the text editor could be used to send things like timesheets, contract information, videos can show progress of work, etc. with the payment request. I can think of more use cases for keeping it, than removing it. The text editor worked perfectly when I tested it :P

britttttk commented:
Kukks I like the improvments you have made. I see you fixed a bug in the pay button in your previous 2435187 commit. So I am going to mention that a bug still exists. I'm sorry if you are not ready for re-review. However, it exists like this:

----

**Example 1**:
- check box to allow payee to make the custom payment
- Payer chooses value, (in the Ex1 below it's 1 BTC) and they pay that value (1 BTC) in full.
- When they try to pay the remaining balance (in this case 9 BTC) it does not allow payment when the user clicks the pay button because it opens the Paid modal invoice from the previous payment. 

![cantpayt8](https://user-images.githubusercontent.com/39231115/52396354-dfbd3900-2a6e-11e9-8509-75bd29030889.PNG)

----

**Example 2:** (same issue as Ex1)
- User pays 55 BTC of their custom 55 BTC invoice. Ok no problem.
- User pays 1 BTC of their custom 40 BTC invoice. Ok no problem.
- User tries to pay remaining balance of 44 BTC. Opens to a Completed Payment modal invoice. 

![openstocompletedpayment](https://user-images.githubusercontent.com/39231115/52396722-0c258500-2a70-11e9-953d-1f1e8d8dc4fa.PNG)

Kukks commented:
hmm ok, it's because the `/pay` mechanism grabs the last invoice. I will need to add a pay button next to the non-complete invoices and add some additional logic 

Thanks for the test cases

britttttk commented:
No problem. I like the custom pay button but it also seemed to work fine before when the user just underpaid the one request. If that was true, is it worth it to add extra pay buttons per request? It sounds to me like it could confuse the user. But I’m not sure how it was/will be.

Kukks commented:
Fixed, did not add any extra pay links. Instead, if there is an active invoice, you will only be able to pay that one.

britttttk commented:
@Kukks it still happens. It looks so much better though. Also the Tx Id it looks like this on Firefox :/ 

![image](https://user-images.githubusercontent.com/39231115/52518145-67d44780-2c03-11e9-97d9-7ee9fcb196ea.png)

Kukks commented:
fiiiiiiiiixed

britttttk commented:
Works perfectly, thanks for fixing :D

NicolasDorier commented:
Nobody bitching about Kukks code anymore? @pavlenex ?
If so, that will be my turn after I reviewed it in my plane back to tokyo :D

Eskyee commented:
bonne voyage 🛩

pavlenex commented:
I haven't audited the code, but I tested the front end and UX of this PR. In my tests, I haven't found any bugs and all earlier bugs were fixed.

The only UX problem I've found is that the amount field always has 0, that needs to be deleted prior to entering a normal amount. I think that should be fixed.

![0](https://user-images.githubusercontent.com/36959754/52727642-f6efaf80-2fb5-11e9-8a44-cb888e6fda6b.PNG)

I'm also not a fan of the 'actions' in the payment request view, especially on a mobile, it's hard to click. I'm afraid there's little to none to be done there until we improve the UI.

![actions](https://user-images.githubusercontent.com/36959754/52727883-77aeab80-2fb6-11e9-8f5f-f2f39ea257bc.PNG)
![actions2](https://user-images.githubusercontent.com/36959754/52727886-78474200-2fb6-11e9-8c76-3dcbaf46959f.PNG)

My suggestion is that in future we make these requests even more customizable and probably allow drag and drop or color customization that will give more customization options for non-coders without having to use CSS.

So afaik, you can review it now @NicolasDorier

NicolasDorier commented on BTCPayServer/PaymentRequest/PaymentRequestHub.cs:
```diff
@@ -131,14 +132,18 @@ protected override async Task ProcessEvent(object evt, CancellationToken cancell
+                if (updated.Published)
                 {
-                    await InfoUpdated(updated.PaymentRequestId);
+                    await _PaymentRequestService.UpdatePaymentRequestStateIfNeeded(updated.Data);
```
why? The `EditPaymentRequest` should have already updated the payment inside `CreateOrUpdatePaymentRequest`.

Kukks replied:
When you save an item and publish( and was in creating state before), the state gets stuck on pending even if the payments were settled

NicolasDorier replied:
How can payments be settled before publish?

Kukks replied:
By setting the payment back to draft mode or a store owner pays it in create mode.

NicolasDorier replied:
it should not be able to be back in draft mode

NicolasDorier commented on BTCPayServer/Models/PaymentRequestViewModels/ListPaymentRequestsViewModel.cs:
```diff
@@ -80,6 +71,8 @@ public UpdatePaymentRequestViewModel(PaymentRequestData data)
         [Display(Name = "Allow payee to create invoices in their own denomination")]
         public bool AllowCustomPaymentAmounts { get; set; }
+
+        public bool Enabled { get; set; }
```
Can remove

NicolasDorier commented on BTCPayServer/Models/PaymentRequestViewModels/ListPaymentRequestsViewModel.cs:
```diff
@@ -124,6 +107,8 @@ public ViewPaymentRequestViewModel(PaymentRequestData data)
             }
         }
 
+        public bool Enabled { get; set; }
```
Can remove

NicolasDorier commented:
Merged!
